Add rapidity methods

graeme-a-stewart opened the pull request:
Add methods that return the true rapidity of `LorentzVectors (E, p)` these are defined for both cartesian and cylindrical cases as ~~`rap()`~~ `rapidity()`.

Add methods for the square of mass and the square of momentum (`mass2`, `pt2`) as these can be useful for more specialist calculations (but not exported).

graeme-a-stewart committed:
Add rapidity methods

Add methods that return the true rapidity of LorentzVectors (E, p)
these are defined for both cartesian and cylindrical cases as rap()

Add methods for the square of mass and the square of momentum
as these can be useful for more specialist calculations (but not
exported)

Moelf commented on README.md:
````diff
@@ -1,23 +1,32 @@
+
 ```julia
-px, py, pz, energy, pt, eta, phi, mass
+px, py, pz, energy, pt, rap, eta, phi, mass
````
I think we might just want to call this `rapidity`? that's what `ROOT` and scikit-hep `vector` calls it

graeme-a-stewart replied:
Good point - it's a better method name

Moelf commented on src/cartesian.jl:
```diff
@@ -1,6 +1,8 @@
-mass(lv::LorentzVector) = sqrt(dot(lv, lv))
-pt(lv::LorentzVector) = sqrt(muladd(lv.x, lv.x, lv.y^2))
+mass2(lv::LorentzVector) = dot(lv, lv)
+mass(lv::LorentzVector) = mass2(lv) < 0.0 ? sqrt(-mass2(lv)) : sqrt(mass2(lv))
```
is this `negative` mass trick universal? what's ROOT's `Math::LorentzVector` behavior on this

graeme-a-stewart replied:
ROOT adopts this convention for space-like 4-vectors, so that's some precedent

Moelf replied:
I'm trying to think if we should err on the side of not silently produce wrong results

graeme-a-stewart replied:
Depends what one thinks of as being *wrong*. Negative mass is not ambiguous - but it may be surprising. I'm ok with rolling this back.

Moelf replied:
Yeah, I guess I mean if we error, it's trivial for user to do what they want. But not the other way, they may silently get wrong results

graeme-a-stewart replied:
I checked in the Python Vector package and it also conforms to this behaviour:

```python
In [1]: import vector

In [2]: ep_vec = vector.obj(x=8.1, y=2.2, z=3.3, E=4.4)

In [3]: ep_vec.mass
Out[3]: -7.87273777030583
```

And I found I missed a `-` sign on the mass method, oops! Fixed in incoming commit.

graeme-a-stewart committed:
Rename rap to rapidity

This is the same method name as ScikitHEP's Vector package
Add a note that we adopt the ROOT convention of returning a negative
mass for space-like 4-vectors

graeme-a-stewart committed:
Improve code coverage

graeme-a-stewart committed:
Fix mass method for space-like vectors

And add test for same

Moelf commented:
Go ahead and merge this when you're ready!

graeme-a-stewart commented:
Thanks @Moelf - merge early, merge often!

Can we foresee a new release that can be a proper dependency for JetReconstruction.jl?
